src/backend/tcop/zenith_wal_redo.c

```diff
@@ -624,7 +631,16 @@ ApplyRecord(StringInfo input_message)
 	redo_read_buffer_filter = redo_block_filter;
 
 	RmgrTable[record->xl_rmid].rm_redo(reader_state);
-
+	/*
+	 * If there is WAL record with FPI, then no image is provided and so PushPage is not called.
+	 * We need to initialize wal_redo_buffers in this case.
+	 */
+	if (BufferIsInvalid(wal_redo_buffer))
+	{
+		wal_redo_buffer = ReadBufferWithoutRelcache(target_redo_tag.rnode, target_redo_tag.forkNum, target_redo_tag.blockNum, RBM_NORMAL, NULL);
```

Comment:
Shouldn't this be RBM_ZERO? I'm surprised RBM_NORMAL works here at all, the WAL redo process cannot read the page from disk.
```suggestion
		wal_redo_buffer = ReadBufferWithoutRelcache(target_redo_tag.rnode, target_redo_tag.forkNum, target_redo_tag.blockNum, RBM_ZERO, NULL);
```

Reply:
I do not think that we need RBM_ZERO here, because it means that we have WAL redo record associated with particular page, but which doesn't initialize this page (otherwise why this page may be absent in buffers after applying this record?).
Also I think that RBM_ZERO can mask some error in WAL applying.
But I committed your suggestion.